presto-spi/src/main/java/io/prestosql/spi/tracer/Tracer.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.tracer;
+
+import io.prestosql.spi.eventlistener.TracerEvent;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+// Tracer keeps track of Presto execution unit contexts and sends debugging purpose events via event listener manager.
+// Users should not build complex metrics based on the action events.
+public class Tracer
+{
+    private final boolean isEnabled;
+    private final Consumer<TracerEvent> emitter;
+    private final String queryId;
+    private final Optional<String> stageId;
+    private final Optional<String> taskId;
+    private final Optional<String> pipelineId;
+    private final Optional<String> splitId;
+    private final Optional<String> operatorId;
+    private final Optional<String> nodeId;
+
+    private Tracer(Builder builder)
```

Comment:
It's better to have a `build` method inside the Builder class that constructs Tracer object with member variables as parameters.